### ISSUE TITLE
Improving suggested words with alternate characters

### ISSUE THREAD

**Issue by HenryWales** [HugoFara/lwt#136]
This is an advanced feature request.

There are many languages where a similar word (different tense, different number, different grammatical category, etc. but same root) won't have a perfect string match, but instead will have some common character substitutions (a short string of characters replaced by another short string of characters).

For example in English, drink, drunk.

The idea would need a language-specific list of common substitutions and a parsing algorithm that can check substitutions. The substitution would still rank lower than a continuous word, of course.

A difficulty is the lack of an objective measure for the relevance of suggestions. There are some substitutions with very high relevance in some languages, but others are less obvious, so creating a parsing list for a language, tweaking how much a substitution penalizes ranking and measuring the end effect have some degree of subjectivity.

Such a solution also probably require some caching, otherwise loading time for suggested words could be problematic.

I would be ready to help, but I can understand if this is not a priority.

**Comment by HenryWales** [comment]
I thought you had stopped development of LWT. It's great to hear you're still improving it. These changes might indeed cover what I was looking for.

Are you planning on doing a new release in the near future? As there is no clean way to separate multiple words sharing the same set of letters, I have increased some limits for some fields. This makes upgrading to the latest version a bit more work.

**Comment by HugoFara** [comment]
Hi Henry,

Quick update on this and your two questions from earlier.

**Lemmatizer is shipped.** Full reference is now at <https://hugofara.github.io/lwt/reference/lemmatization>. The short version of what should cover #136:

- Strategy is selected per language via the `LgLemmatizerType` column: `none`, `dictionary`, `spacy`, or `hybrid`.
- 24 languages get high-quality spaCy lemmatization out of the box (`en`, `de`, `fr`, `es`, `it`, `pt`, `ru`, `nl`, `pl`, `ja`, `zh`, `ko`, `uk`, `ro`, `da`, `sv`, …).
- For anything spaCy doesn't cover well, or for domain-specific corrections, drop a TSV `{lang}_lemmas.tsv` into `data/lemma-dictionaries/` and the dictionary lemmatizer picks it up. That's the language-specific substitution list you originally asked for.
- `hybrid` is dictionary first, spaCy fallback. That's the recommended default when both are available.
- Lemmas are persisted on each term (`words.WoLemma` / `WoLemmaLC`) and the spaCy models are cached on the NLP service side, so there's no per-request NLP cost after the first lookup. It addresses the caching concern you raised.

**Release / testing.** v3 has been out for a while; the latest tag is 3.1.0-fork. The official public instance is not here yet though (no time on my side), but you can easily test with Docker.

**Field limits.** Tangential to this issue, but if you can share which columns you had to extend in your fork, I'll check whether the upstream limits should follow.

Could you confirm two things so I can close this?

1. Does the spaCy + dictionary combination cover your original drink/drunk-style use case well enough?
2. Anything still missing that you'd want to track in a narrower follow-up issue?

**Comment by HugoFara** [comment]
Done in `7f5a46fe3`, with a different implementation.

Terms sharing the searched term's **lemma** are now suggested first, whatever they score on spelling. That covers what a substitution list would (`drink` /`drunk`) plus what it couldn't reach without an entry per exception (`bought`/`buy`, `feet`/`foot`, `worse`/`bad`, all of which score exactly 0.000 on letter pairs). 

On the caching concern: a saved term reuses the lemma stored when you saved it, and an unsaved one is looked up in the local dictionary only, never the NLP service, so the box never waits on the network.

Caveat: the bundled `en_lemmas.tsv` is a 142-entry sample. If your language has spaCy or a custom TSV configured you get the full benefit; otherwise the win is small until we ship real dictionaries, which I'll track separately.

Closing this, please reopen if it doesn't cover your case.
